rfcs/rfc-3-layout-callbacks.md

```diff
@@ -0,0 +1,222 @@
+
+1. It's **fragile** - We only receive the last calculated layout and may need to force a render with `context.mark_dirty()` until we don't detect any changes.
+2. It's **dumb** - Since we only force a re-render when the layout is different, layouts that take more than a frame to calculate are not caught (specifically, surrounding our content in an `Element` with width and height set to `Units::Auto` will not be calculated correctly).
+
+Because of this, I think we need a better solution for getting layout information.
+
+## Problem
+
+There's currently no proper way to respond to a widget layout. We either need to just hope it's fully calculated within a set number of renders, set it manually, or perform some other hack. Instead, we need some way to properly be notified of a widget's layout.
+
+## Solution
+
+There's a few ways to go about this, but one of the simplest will be to add an `on_layout` event callback in manner similar to React Native's [`onLayout`](https://reactnative.dev/docs/view#onlayout) prop.
+
+Additionally, morphorm exposes a [`GeometryChanged`](https://github.com/geom3trik/morphorm/blob/cd4032e4409e913b4b4f2f7424761530d1652970/src/types.rs#L98-L119) bitflag struct that could be passed along to the end user so they know if anything changed or *will* change.
+
+### Desired API
+
+The desired API would then look something like:
+
+```rust
+#[widget]
+fn MyWidget() {
+  let on_layout = OnLayout::new(|ctx, event| {
+    // --- Get Current Layout --- //
+    let width = event.layout.width;
+    let height = event.layout.height;
+    let x = event.layout.x;
+    let y = event.layout.y;
+    let z = event.layout.z;
+    
+    // --- Get Change Flags --- //
+    let pos_change = GeometryChanged::POSX_CHANGED | GeometryChanged::POSY_CHANGED;
+    if(pos_change == event.flags) {
+      println!("Position changed!");
+    }
+    
+    // --- Do Something With Widget --- //
+    let widget_id = event.target;
+    // ...
+    
+  });
+  
+  rsx! {
+    <Element on_layout={Some(on_layout)} />
+  }
+}
+```
+
+### Data Types
+
+At a basic level, this API would need the following types: `Layout`, `LayoutEvent`, and `OnLayout`.
+
+#### `Layout`
+
+While we could just use morphorm's `Rect` type here, it might be best to create our own `Layout` type to replace it. This allows us to define our own methods and impls on it in ways we couldn't do with morphorm. It also means that a breaking change in morphorm's `Rect` wouldn't affect our end-users.
+
+```rust
+pub struct Layout {
+  pub width: f32,
+  pub height: f32,
+  pub x: f32,
+  pub y: f32,
+  pub z: f32,
+}
+
+impl Layout {
+  /// Returns the position as a Kayak position type (which is currently just a tuple)
+  pub fn pos(&self) -> (f32, f32) {/* ... */}
+}
+
+impl Into<Rect> for Layout {/* ... */}
+impl From<Rect> for Layout {/* ... */}
```

Comment:
I just saw that `Rect` is actually a type from Kayak, not morphorm. Do we want to rename that struct and its fields to match `Layout`'s? Or do we want to keep them separate?

Reply:
Lets keep them separate for now.